python/ray/serve/_private/controller.py

```diff
@@ -1032,6 +1046,63 @@ def _get_logging_config(self) -> Tuple:
         return self.system_logging_config, log_file_path
 
 
+def calculate_scale_direction(
+    new_config: ServeDeploySchema,
+    all_app_statuses: Dict[str, ApplicationStatusInfo],
+    curr_target_capacity: Optional[float],
+    curr_scale_direction: Optional[float],
+) -> Optional[TargetCapacityScaleDirection]:
+    next_scale_dirction = None
+
+    if len(all_app_statuses) == 0 and new_config.target_capacity is not None:
+        next_scale_dirction = TargetCapacityScaleDirection.UP
+
+    if live_applications_match_config(new_config, all_app_statuses):
+        next_target_capacity = new_config.target_capacity
+
+        if curr_target_capacity == next_target_capacity:
+            next_scale_dirction = curr_scale_direction
+        elif curr_target_capacity is None and next_target_capacity is not None:
+            next_scale_dirction = TargetCapacityScaleDirection.DOWN
+        elif curr_target_capacity is None and next_target_capacity is not None:
+            next_scale_dirction = None
+        elif curr_target_capacity < next_target_capacity:
+            next_scale_dirction = TargetCapacityScaleDirection.UP
+        else:
+            assert curr_target_capacity > next_target_capacity
+            next_scale_dirction = TargetCapacityScaleDirection.DOWN
+    else:
+        next_scale_dirction = curr_scale_direction
+
+    if next_scale_dirction != curr_scale_direction:
+        if isinstance(next_scale_dirction, TargetCapacityScaleDirection):
+            logger.info(
+                f"Target capacity scaling {next_scale_dirction.value.lower()} "
+                f"from {curr_target_capacity} to {new_config.target_capacity}."
+            )
+        else:
+            assert next_scale_dirction is None
+            logger.info(f"Target capacity entering 100% at steady state.")  # noqa: F541
+
+    return next_scale_dirction
+
+
+def live_applications_match_config(
+    config: ServeDeploySchema, all_app_statuses: Dict[str, ApplicationStatusInfo]
+) -> bool:
+    config_apps = {app.name for app in config.applications}
+
+    live_apps = set()
+    for (
+        app_name,
+        status_info,
+    ) in all_app_statuses.items():
+        if status_info.status != ApplicationStatus.DELETING:
```

Comment:
Yes, that's intentional. I needed some way to define whether the new config is the "same" or "different", so Serve could tell when to use the previous config to determine the scaling direction.

If the previous config is the "same" as the new config, then Serve can compare the previous config's `target_capacity` with the new config's `target_capacity` to determine the scaling direction.

If the previous config is "different" than the new config, then Serve shouldn't use the previous config's `target_capacity`.

I define two configs to be the "same" if they both contain apps of the same name. Since Serve doesn't explicitly track configs, I check whether the new config is the same as previous config by comparing the new config's apps with the previous config's apps. All active apps, except the ones that are `DELETING`, must be from the previous config. That's why I call them "live apps"– because they're currently live on the cluster. `DELETING` apps are from configs before the previous config; they're not live anymore, and they will be removed.

Reply:
I just realized that the controller already reconstructs the previous config in `_recover_config_from_checkpoint`. I could use that same logic to compare the configs directly. I'll make the changes later today.

Reply:
I changed this method to compare configs directly. Let me know what you think!

Reply:
LGTM! Much more clear on the intention!